### PR TITLE
chore(observability): Internal tracing RFC

### PR DESCRIPTION
**Pull request by lukesteensen**
[Rendered](https://github.com/timberio/vector/blob/tracing-rfc/rfcs/2021-08-13-8025-internal-tracing.md)

### CODE REVIEW

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+### Implementation
+
+#### Overview
+
+Vector already has some very basic trace-style instrumentation via the
+[`tracing`](https://docs.rs/tracing/) crate that we currently use primary for
+logging. This RFC proposes that we expand that instrumentation to include more
+granular spans around the existing events (i.e. `info!` and friends, not Vector
+events).
```
Comment (binarylogic) [inline rfcs/2021-08-13-8025-internal-tracing.md:49]: Vector events call `info!`, correct? In other words, all of Vector's event-driven log lines will be included?

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+The ongoing maintenance burden should be quite minimal. The extent of code
+changes would be the addition of some small span instantiations and
+instrumentation, but nothing that would require threading through existing
+interfaces or reorganizing any abstractions.
+
+## Prior Art
+
+There is very little prior art available for tracing streaming system, but we
+are largely just using existing off-the-shelf components. We've not opted to
+invent anything particularly new for Vector.
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:198]: One thing that had come up in my profiling reading struck a chord here: user-space dtrace probes. Downsides with these are there's no support for some of our deploy targets -- Windows -- but we avoid any runtime impact that tracing currently imparts. Oxide is using them in their work: https://github.com/oxidecomputer/usdt

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+- What is an acceptable performance overhead for enabling tracing in user environments?
+- Are component-focused spans acceptable, or are there cases where event-focused spans would be clearly superior?
+- Are we comfortable with the proposed methods of avoiding the previous memory growth issue?
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues
+after the RFC is approved:
+
+- [ ] Adjust existing spike-level implementation to fully match what's described here
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:240]: I have had some questions about tracing's runtime impact. If we could contrive a feature flag that let us turn off tracing that'd be valuable for getting good performance impact numbers.

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+intended.
+
+In summary:
+
+- Spans should be component-focused, breaking down their operations by stages with at least two levels of hierarchy but up to three levels for components with more complex internals:
+    1. Single "iteration" of operation, on the order of seconds, as a baseline
+    2. High level stages of operation within an iteration, such as reading,
+       processing, sending
+    3. Interesting sub-stages of each operation, with a focus on units of work
+       with significant off-CPU time
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:124]: The view we have now from metrics and profiling is aggregate, in that it's done across multiple `Event`s flowing through vector. The value I would see from tracing, especially, would be something that's per-`Event`, so if we have an expensive path through vector we can actually see that path in practice. Is the proposal here more like our existing observation methods -- aggregate --  or do we get a more concrete view? That's where I'd see real value, being able to trace the flow of an `Event` through vector, especially considering the runtime impact.

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+## Rationale
+
+### Why is this change worth it?
+
+With a relatively small investment of effort, the system described here would be
+able to deliver significant, Vector-specific insight into performance directly
+from deployed environments. The type of instrumentation needed is already
+present within Vector, just underutilized, and it very neatly fills a gap
+between high-level metrics and low-level CPU profiling.
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:144]: Is there more of this gap we can fill without tracing? IIRC @jszwedko is tracking where we could have more metrics present in the system and I have thoughts as well from the performance work. There's a tension between the need for better observability into vector's operation and its throughput performance. I guess I want to be real sure we've wrung out all we can from cheaper methods before we start adding in more work that regresses throughput. 

I bring this up because the more I see tracing in our profiling output without getting real benefit from it the more tempted I am to ax it from the codebase.

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+How we instrument Vector components with spans should be driven by the kinds of
+questions we'll be looking to answer with the resulting data. In particular, the
+focus should be on the types of questions that are difficult to answer with our
+current tools (metrics, logs, and sometimes Datadog's native profiler).
+
+A typical performance investigation starts at a very high level and works its
+way down into specifics. The first step tends to be identifying which component
+is the bottleneck. This is often done with some intuition and/or experimentation
+(e.g. "Lua tends to be slower, try dropping it and see what happens"), and the
+new utilization metrics are designed to help with exactly this.
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:65]: I agree that this is the loop. I'd like to see more discussion of how adding in tracing would shorten this loop. The implication is that we'd see -- from customer environment? -- that such and such internal piece of a component is slow and be able to focus our efforts on that. If so, this gets into the same brand of issue that CPU profile does, how to distinguish a time hog from an actual bottleneck. Getting tracing into customer environments is a step beyond just getting into soak environments, so perhaps motivation from one of the [investigation tickets](https://github.com/timberio/vector/issues/8263) would valuable. Questions I have are, how would tracing as suggested here have shortened the work in the linked investigation, especially compared to the results we got from CPU profiling / metrics.

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+present within Vector, just underutilized, and it very neatly fills a gap
+between high-level metrics and low-level CPU profiling.
+
+### What is the impact of not doing this?
+
+Performance investigations would continue to rely on a combination of intuition
+and expert-level tooling to narrow down problems reported by users. Instead of
+having data directly indicating that, for example, a component is spending all
+of its time fingerprinting files, we would need to deduce that from other
+sources of information.
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:152]: How would this experience differ -- where we request users to send us traces -- from requesting users send us sampled CPU profiles? Both point to where vector was a time hog and require some knowledge of vector internals to interpret / act on.

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+Incremental steps to execute this change. These will be converted to issues
+after the RFC is approved:
+
+- [ ] Adjust existing spike-level implementation to fully match what's described here
+- [ ] Instrument enough components to fully cover a chosen soak test scenario
+- [ ] Enable tracing in that soak test scenario, gathering data on performance impact
+- [ ] Open issues to add spans to the remaining components, organize and prioritize implementation
+
+Note: The final step is not necessarily required, and we could take the stance
+that this kind of instrumentation can happen as it is needed.
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:246]: I don't want to presume we're moving forward with tracing in vector. Biggest concerns are:

* quantifying the runtime impact
* avoiding run-away memory growth accidentally
* sufficient differentiation from metrics/CPU profiling to be worth the costs

We do need to run in soaks to get some sense of this but I don't see a bail out point in the plan of attack here. I argue we should constrain this RFC some. Customer end-goals are great but maybe we just focus on going live in a single soak test for now, `datadog_agent -> remap -> datadog_logs`. Can we quantify 

* the performance impact of getting proposed traces from this setup
* the performance impact of excising tracing entirely from this setup
* the performance impact of tracing being present but latent versus flipped on and
* how this compares in practice to CPU profiling?

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+## Future Improvements
+
+- Document specific performance impact of tracing based on measurements in our soak testing infrastructure
+- Improve observability of trace shipping infrastructure itself
+- Add more granular but more expensive span data at log levels like `DEBUG`
+- Enable sampling and options helping us towards a potential always-on system
+- Remove requirement of running a local Datadog agent
```
Comment (blt) [inline rfcs/2021-08-13-8025-internal-tracing.md:254]: As above I think this presumes that we're moving forward with tracing. FWIW my appetite is for a bounded experiment to determine how much of a game changer tracing is compared to our existing observability methods -- especially if we improve those -- to help decide whether we ought to move forward with it or pare back / remove it from the codebase. This RFC does mention that we're slightly outside of the normal request/response model for tracing.

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 5 above)
Comment (lukesteensen) [reply rfcs/2021-08-13-8025-internal-tracing.md:144]: @blt 

> the more I see tracing in our profiling output without getting real benefit from it the more tempted I am to ax it from the codebase.

Could you be a little more specific here? The unfortunate collision between "tracing" the concept (discussed in this RFC, involving adding more spans and plumbing to ship them) and `tracing` the crate (as would show in current profiling output, used for logging and metrics context) makes this a little hard to parse. 

Axing the idea of shipping span data would be very simple, but axing `tracing` would require finding replacements for both internal logging and tagging of internal metrics. I've been a reasonably happy user of `tracing` so far, but if we have specific performance concerns around our current use then those should be resolved prior to further investment of the kind discussed in this RFC.

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 3 above)
Comment (lukesteensen) [reply rfcs/2021-08-13-8025-internal-tracing.md:240]: AFAIK, `tracing` is designed such that events and spans with no interested subscribers are never constructed. This effectively means that by turning the log level up to `CRITICAL` or similar, very little should actually happen at `span` and `event` callsites. That's obviously not the same as turning it off entirely, but in practice it's intended to be very close.

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+(e.g. "Lua tends to be slower, try dropping it and see what happens"), and the
+new utilization metrics are designed to help with exactly this.
+
+The step after that is a bit fuzzy, however. You could jump straight to
+profiling, but it has a few weaknesses:
+
+1. It's not something that can easily be run in customer environments
+1. The focus on CPU time vs wall clock time means it can miss issues like bad
+   rate limits, waiting on downstream components, slow IO operations or
+   syscalls, etc
```
[the remark this answers is not in the document]
Comment (lukesteensen) [reply rfcs/2021-08-13-8025-internal-tracing.md:73]: Yes, that's a good point. We certainly have not reached the limit of useful metrics we could be emitting, and it's possible that we could achieve much of the same benefit as spans by just shipping more histograms.

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 4 above)
Comment (lukesteensen) [reply rfcs/2021-08-13-8025-internal-tracing.md:124]: The tracing described here would not necessarily be per-`Event`, but I'm not sure that it'd really qualify as aggregate either. I'd also question the premise that per-`Event` is more "concrete" or valuable.

The goal is to provide more insight into how Vector is working internally, and Vector's internal operation is not necessarily focused on individual `Event`s. In the file source, for example, many of the issues we've fixed have occurred entirely before an `Event` is even constructed. In a transform, on the other hand, spans would likely cover the processing of a single event (or hopefully batch of events).

So the focus of this RFC is on tracing "operations", and it allows that to be defined however is most convenient for a given component. There's a bit of an implicit assumption that individual events are unlikely to be particularly interesting, at least relative to the degree by which some web requests can be more interesting than others, and that Vector has fewer, less variable paths of execution than your average web server. This could be taken as an argument that tracing is a less valuable tool for Vector.

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 1 above)
Comment (lukesteensen) [reply rfcs/2021-08-13-8025-internal-tracing.md:49]: There are three things called events here:

1. Vector's `Event`: the actual user data flowing through the system
2. Vector's `internal_events`: the abstraction we built for emitting internal observability data
3. The `tracing` crate's `Event`: the data created with `info!` and friends, that manifest as Vector's own logging

We're not talking about shipping any of these as part of tracing, just the spans that surround and give context to (3).

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 5 above)
Comment (blt) [reply rfcs/2021-08-13-8025-internal-tracing.md:144]: Oh yeah, span data. I guess it is kind of an overloaded thing in our project. :) 

For instance, [this change](https://github.com/timberio/vector/pull/8579) ended up being worth +9Mb/s. The question for every span is then can we avoid its overhead or is the span providing information that's worth the overhead? We can maybe bound this by removing all spans and measuring vector's performance span-less.

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 4 above)
Comment (blt) [reply rfcs/2021-08-13-8025-internal-tracing.md:124]: Makes sense.

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 3 above)
Comment (blt) [reply rfcs/2021-08-13-8025-internal-tracing.md:240]: Oh yeah, that'd be an interesting experiment and might let us cheaply get at what a "no span" vector looks like, with regard to throughput.

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 5 above)
Comment (lukesteensen) [reply rfcs/2021-08-13-8025-internal-tracing.md:144]: Gotcha, that makes sense. We have had at least one other major performance degradation due to unleveled spans, so it's definitely a concern.

File: rfcs/2021-08-13-8025-internal-tracing.md [hunk 11]
```diff
@@ -4,6 +4,19 @@ This RFC discusses an integration with the Datadog APM product for Vector's
 internal tracing data, with the goal of enabling faster and more accurate
 diagnosis of performance issues in deployed environments.
 
+## Status
+
+The team decided not to move forward with this change. Implementing tracing
+within Vector is not a worthy goal in itself, and there is a clearer, cheaper
+path to addressing the pain points laid out in this RFC via more granular timing
+metrics.
```
Comment (bruceg) [inline rfcs/2021-08-13-8025-internal-tracing.md:12]: Is it worth noting the eBPF discussion as well?

**Review by blt — APPROVED** [review]

File: rfcs/2021-08-13-8025-internal-tracing.md (hunk 11 above)
Comment (lukesteensen) [reply rfcs/2021-08-13-8025-internal-tracing.md:12]: I've opened a confluence page where some more detailed discussion can live. Not linked here because it's internal.

**Review by jszwedko — APPROVED** [review]
